frame/dapps-staking/src/testing_utils.rs

```diff
@@ -394,43 +394,93 @@ pub(crate) fn assert_withdraw_unbonded(staker: AccountId) {
 
     // Calculate contract portion of the reward
-    let (_, stakers_joint_reward) =
-        DappsStaking::dev_stakers_split(&init_state.contract_info, &init_state.era_info);
+    let (_, stakers_joint_reward) = DappsStaking::dev_stakers_split(
+        &init_state_claim_era.contract_info,
+        &init_state_claim_era.era_info,
+    );
 
-    let (claim_era, staked) = init_state.staker_info.clone().claim();
+    let (claim_era, staked) = init_state_claim_era.staker_info.clone().claim();
     assert!(claim_era > 0); // Sanity check - if this fails, method is being used incorrectly
 
     // Cannot claim rewards post unregister era, this indicates a bug!
-    if let DAppState::Unregistered(unregistered_era) = init_state.dapp_info.state {
+    if let DAppState::Unregistered(unregistered_era) = init_state_claim_era.dapp_info.state {
         assert!(unregistered_era > claim_era);
     }
 
     let calculated_reward =
-        Perbill::from_rational(staked, init_state.contract_info.total) * stakers_joint_reward;
+        Perbill::from_rational(staked, init_state_claim_era.contract_info.total)
+            * stakers_joint_reward;
     let issuance_before_claim = <TestRuntime as Config>::Currency::total_issuance();
 
     assert_ok!(DappsStaking::claim_staker(
         Origin::signed(claimer),
         contract_id.clone(),
     ));
+
+    let final_state_claim_era = MemorySnapshot::all(claim_era, contract_id, claimer);
+    let final_state_current_era = MemorySnapshot::all(current_era, contract_id, claimer);
+
+    // If there's less than 2 events, it should fail anyway, so panic as acceptable
+    let events = dapps_staking_events();
+    let second_last_event = &events[events.len() - 2];
+
+    if DappsStaking::should_restake_reward(
+        init_state_claim_era.ledger.reward_destination,
+        contract_id,
+        init_state_claim_era.staker_info.latest_staked_value(),
+    ) {
+        assert_eq!(
+            second_last_event.clone(),
+            Event::<TestRuntime>::RewardAndRestake(claimer, contract_id.clone(), calculated_reward,)
+        );
+        assert_eq!(
+            init_state_claim_era.staker_info.latest_staked_value() + calculated_reward,
+            final_state_claim_era.staker_info.latest_staked_value()
+        );
+        assert_eq!(
+            init_state_claim_era.ledger.locked + calculated_reward,
+            final_state_claim_era.ledger.locked
+        );
+
+        assert_eq!(
+            init_state_current_era.era_info.staked + calculated_reward,
+            final_state_current_era.era_info.staked
+        );
+        assert_eq!(
+            init_state_current_era.era_info.locked + calculated_reward,
+            final_state_current_era.era_info.locked
+        );
+    } else {
+        assert_eq!(
+            init_state_claim_era.free_balance + calculated_reward,
+            final_state_claim_era.free_balance
+        );
```

Comment:
Isn't this check also applicable if `if` clause is `true`?

Reply:
yes, it's a mistake, it should check free balance in the `current_era` is higher after claim without compounding